### PR TITLE
Clarify docs for when binary_search has many matches.

### PR DESCRIPTION
**Pull request by frewsxcv**
Fixes https://github.com/rust-lang/rust/issues/51817.

### CODE REVIEW

File: src/libcore/slice/mod.rs [hunk 1]
```diff
@@ -1174,10 +1174,11 @@ impl<T> [T] {
 
     /// Binary searches this sorted slice for a given element.
     ///
-    /// If the value is found then `Ok` is returned, containing the
-    /// index of the matching element; if the value is not found then
-    /// `Err` is returned, containing the index where a matching
-    /// element could be inserted while maintaining sorted order.
+    /// If the value is found then `Ok` is returned, containing the index of
```
Comment (GuillaumeGomez) [inline src/libcore/slice/mod.rs]: Please add link to variant.

File: src/libcore/slice/mod.rs [hunk 2]
```diff
@@ -1174,10 +1174,11 @@ impl<T> [T] {
...
+    /// the matching element. If there are multiple matches, then any one of the
+    /// matches could be returned. If the value is not found then `Err` is
```
Comment (GuillaumeGomez) [inline src/libcore/slice/mod.rs]: Please add link to variant.

File: src/libcore/slice/mod.rs [hunk 3]
```diff
@@ -1208,10 +1209,11 @@ impl<T> [T] {
     /// order code that indicates whether its argument is `Less`,
     /// `Equal` or `Greater` the desired target.
     ///
-    /// If a matching value is found then returns `Ok`, containing
-    /// the index for the matched element; if no match is found then
-    /// `Err` is returned, containing the index where a matching
-    /// element could be inserted while maintaining sorted order.
+    /// If the value is found then `Ok` is returned, containing the index of
```
Comment (GuillaumeGomez) [inline src/libcore/slice/mod.rs]: Please add link to variant.

File: src/libcore/slice/mod.rs [hunk 4]
```diff
@@ -1208,10 +1209,11 @@ impl<T> [T] {
...
+    /// the matching element. If there are multiple matches, then any one of the
+    /// matches could be returned. If the value is not found then `Err` is
```
Comment (GuillaumeGomez) [inline src/libcore/slice/mod.rs]: Please add link to variant.

File: src/libcore/slice/mod.rs [hunk 5]
```diff
@@ -1264,10 +1266,11 @@ impl<T> [T] {
     /// Assumes that the slice is sorted by the key, for instance with
     /// [`sort_by_key`] using the same key extraction function.
     ///
-    /// If a matching value is found then returns `Ok`, containing the
-    /// index for the matched element; if no match is found then `Err`
-    /// is returned, containing the index where a matching element could
-    /// be inserted while maintaining sorted order.
+    /// If the value is found then `Ok` is returned, containing the index of
```
Comment (GuillaumeGomez) [inline src/libcore/slice/mod.rs]: Please add link to variant.

File: src/libcore/slice/mod.rs [hunk 6]
```diff
@@ -1264,10 +1266,11 @@ impl<T> [T] {
...
+    /// the matching element. If there are multiple matches, then any one of the
+    /// matches could be returned. If the value is not found then `Err` is
```
Comment (GuillaumeGomez) [inline src/libcore/slice/mod.rs]: Please add link to variant.

**Comment by frewsxcv** [comment]
@GuillaumeGomez Any suggestions how I should add links when these docs get used on `slice` and `Vec` pages?

File: src/libcore/slice/mod.rs [hunk 7]
```diff
@@ -1174,10 +1174,14 @@ impl<T> [T] {
 
     /// Binary searches this sorted slice for a given element.
     ///
-    /// If the value is found then `Ok` is returned, containing the
-    /// index of the matching element; if the value is not found then
-    /// `Err` is returned, containing the index where a matching
-    /// element could be inserted while maintaining sorted order.
+    /// If the value is found then [`Ok`] is returned, containing the index of
```
Comment (GuillaumeGomez) [inline src/libcore/slice/mod.rs]: Result::Ok should word.

File: src/libcore/slice/mod.rs [hunk 8]
```diff
@@ -1174,10 +1174,14 @@ impl<T> [T] {
...
+    /// the matching element. If there are multiple matches, then any one of the
+    /// matches could be returned. If the value is not found then [`Err`] is
```
Comment (GuillaumeGomez) [inline src/libcore/slice/mod.rs]: Result::Err should word.

File: src/libcore/slice/mod.rs [hunk 9]
```diff
@@ -1174,10 +1174,14 @@ impl<T> [T] {
...
+    /// returned, containing the index where a matching element could be
+    /// inserted while maintaining sorted order.
+    ///
+    /// [`Err`]: result/enum.Result.html#variant.Err
```
Comment (GuillaumeGomez) [inline src/libcore/slice/mod.rs]: Remove those two lines.

**Comment by frewsxcv** [comment]
@GuillaumeGomez CI is passing. r?

**Comment by GuillaumeGomez** [comment]
Thanks!

@bors: r+ rollup

**Pull request merged**